Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +20,88 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+return totalAmpHoursUsed_;
+
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    return isCharging_;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+   return timeWhenChargedOrDepleted_;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
+    //Q_UNUSED(batteryData);
     // This is where you can update your variables
     // Hint: There are many different ways that the totalAmpHoursUsed can be updated
     // i.e: Taking a running average of your data values, using most recent data points, etc.
+
+    old_current=present_batterycurrent;
+    present_batterycurrent = batteryData.current;
+    old_time=present_time;
+    present_time= batteryData.time;
+
+    //total AmpHours Used
+    double initial_amountof_amphours_used =BATTERY_AMP_HOUR_CAPACITY*initialStateOfChargePercent_/100;
+    double average_current= (present_batterycurrent+old_current)/2;
+
+
+    if (old_time.isValid())
+    {
+     double time_difference_in_hours= old_time.msecsTo(present_time)/3600000.0;// For some reason this comes out to 0
+     //QString s= old_time.toString("hh:mm:ss:zzz");
+     //QString t= present_time.toString("hh:mm:ss:zzz");
+     //QTextStream(stdout)<<s<<"      "<<t<<"     "<< time_difference_in_hours<<"     "<< average_current<< "    ";
+    totalAmpHoursUsed_+= average_current*time_difference_in_hours;
+    }
+    else
+        totalAmpHoursUsed_=initial_amountof_amphours_used;
```

Comment:
Comments and spacing. Also bracket the else statement if you bracket the if statement.